### PR TITLE
Create soil layer

### PR DESCRIPTION
**Pull request by aronnoordam**

### CODE REVIEW

File: stem/geometry.py [hunk 1]
```diff
@@ -273,16 +342,17 @@ def __create_surface(geo_data: Dict[str, Any], surface_id: int):
...
         """
 
         # Initialise point and line lists
         points = []
         lines = []
 
         # create surface and lower dimensional objects
-        surface = Surface(abs(surface_id))
-        surface.line_ids = geo_data["surfaces"][surface.id]
+        surface_id = abs(surface_id)
```
Comment (brunozc) [inline stem/geometry.py:354]: why abs?
if you want to enforce being positive we should add a check.

File: stem/model.py [hunk 2]
```diff
@@ -1,24 +1,164 @@
...
     A class to represent the main model.
 
     Attributes:
+        - ndim (int): Number of dimensions of the model
         - project_parameters (dict): A dictionary containing the project parameters.
         - solver (Solver): The solver used to solve the problem.
-        - body_model_parts (list): A list containing the body model parts.
-        - process_model_parts (list): A list containing the process model parts.
+        - geometry (Optional[:class:`stem.geometry.Geometry`]) The geometry of the whole model.
+        - body_model_parts (List[BodyModelPart]): A list containing the body model parts.
```
Comment (brunozc) [inline stem/model.py:20]: missing `

Comment (brunozc) [inline stem/model.py:20]: missing `

File: stem/model.py [hunk 3]
```diff
@@ -1,24 +1,164 @@
...
+        - process_model_parts (List[ModelPart]): A list containing the process model parts.
+        - extrusion_length(Optional[Sequence[float]]): The extrusion length in x,y and z direction
```
Comment (brunozc) [inline stem/model.py:22]: - extrusion_length (Optional[Sequence[float]]): The extrusion length in x, y and z direction

File: stem/model.py [hunk 4]
```diff
@@ -1,24 +1,164 @@
...
+
+    def add_soil_layer_by_coordinates(self, coordinates: Sequence[Sequence[float]],
+                       material_parameters: Union[SoilMaterial, StructuralMaterial], name: str,
+                       ):
+        """
+        Adds a soil layer to the model by giving a sequence of 2D coordinates. In 3D the 2D geometry is extruded in
+        the direction of the extrusion_length
+
+        Args:
+            - coordinates (Sequence[Sequence[float]]): The coordinates of the soil layer.
```
Comment (brunozc) [inline stem/model.py:103]: - coordinates (Sequence[Sequence[float]]): The 2D coordinates of the soil layer.

File: stem/model_part.py [hunk 5]
```diff
@@ -3,23 +3,24 @@
...
+
 
 class ModelPart:
     """
     One part of the complete model, this can be a boundary condition, a loading or another special process
     like excavation.
 
     Attributes:
-        - name (str): name of the model part
+        - name (Optional[str]): name of the model part
```
Comment (brunozc) [inline stem/model_part.py:15]: optional term should be at the end

File: stem/model_part.py [hunk 6]
```diff
@@ -3,23 +3,24 @@
...
         - nodes (np.array or None): node id followed by node coordinates in an array
         - elements (np.array or None): element id followed by connectivities in an array
         - conditions (np.array or None): condition id followed by connectivities in an array
         - geometry (Optional[:class:`stem.geometry.Geometry`]): geometry of the model part
         - parameters (dict): dictionary containing the model part parameters
     """
     def __init__(self):
-        self.name = None
+        self.name: Optional[str] = None
```
Comment (brunozc) [inline stem/model_part.py:23]: why type int for the name and not for the others?

File: tests/test_model.py [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+        for generated_line, expected_line in zip(generated_geometry.lines, expected_geometry.lines):
+            assert generated_line.id == expected_line.id
+            assert generated_line.point_ids == expected_line.point_ids
+
+        # check if surfaces are added correctly
+        for generated_surface, expected_surface in zip(generated_geometry.surfaces, expected_geometry.surfaces):
+            assert generated_surface.id == expected_surface.id
+            assert generated_surface.line_ids == expected_surface.line_ids
+
+    def test_add_multiple_soil_layers_2D(self, expected_geometry_two_layers_2D: Tuple[Geometry, Geometry],
```
Comment (brunozc) [inline tests/test_model.py:303]: should also add a 3D test

File: tests/test_model.py [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+                               Line.create([6, 7], 7),
+                               Line.create([7, 3], 8)]
+
+        full_geometry.surfaces = [Surface.create([1, 2, 3, 4, 5], 1),
+                                  Surface.create([3, 6, 7, 8], 2)]
+
+        return geometry_1, geometry_2, full_geometry
+
+    @pytest.fixture
+    def create_default_2d_soil_material(self):
```
Comment (brunozc) [inline tests/test_model.py:163]: should also add a 3D test

File: tests/test_model.py [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+                            Line.create([6, 3], 7)]
+
+        geometry_2.surfaces = [Surface.create([3, 5, 6, 7], 2)]
+
+        geometry_2.volumes = []
+
+        return geometry_1, geometry_2
+
+    @pytest.fixture
+    def expected_geometry_two_layers_2D_after_sync(self):
```
Comment (brunozc) [inline tests/test_model.py:94]: should also add a 3D test

File: tests/test_model.py [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+            for generated_line, expected_line in zip(generated_geometry.lines, expected_geometry.lines):
+                assert generated_line.id == expected_line.id
+                assert generated_line.point_ids == expected_line.point_ids
+
+            # check if surfaces are added correctly
+            for generated_surface, expected_surface in zip(generated_geometry.surfaces, expected_geometry.surfaces):
+                assert generated_surface.id == expected_surface.id
+                assert generated_surface.line_ids == expected_surface.line_ids
+
+    def test_synchronise_geometry(self, expected_geometry_two_layers_2D_after_sync: Tuple[Geometry, Geometry],
```
Comment (brunozc) [inline tests/test_model.py:412]: should also add a 3D test

File: tests/test_model.py [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+            for generated_line, expected_line in zip(generated_geometry.lines, expected_geometry.lines):
+                assert generated_line.id == expected_line.id
+                assert generated_line.point_ids == expected_line.point_ids
+
+            # check if surfaces are added correctly
+            for generated_surface, expected_surface in zip(generated_geometry.surfaces, expected_geometry.surfaces):
+                assert generated_surface.id == expected_surface.id
+                assert generated_surface.line_ids == expected_surface.line_ids
+
+    def test_add_all_layers_from_geo_file(self, expected_geometry_two_layers_3D: Tuple[Geometry, Geometry]):
```
Comment (brunozc) [inline tests/test_model.py:362]: should also add a 2D test

**Review by brunozc — CHANGES_REQUESTED** [review]

File: stem/geometry.py [hunk 12]
```diff
@@ -250,15 +319,15 @@ def __set_line(geo_data: Dict[str,Any], line_id: int):
...
+            - Tuple[:class:`Line`, Sequence[:class:`Point`]]: The line object and the points that make up the line.
         """
 
         # Initialise point list
         points = []
 
         # create line and lower dimensional objects
-        line = Line(abs(line_id))
-        line.point_ids = geo_data["lines"][line.id]
+        line_id = abs(line_id)
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/geometry.py:329]: gmsh returns negative ids in case the element is inverted, but i dont want that in stem. Same for the rest

**Pull request merged**